parsec/api/protocol/base.py

```diff
@@ -167,3 +237,238 @@ def timestamp_out_of_ballpark_rep_dump(
                 "client_timestamp": client_timestamp,
             }
         )
+
+
+class CmdReqMeta(type):
+    BASE_SCHEMA_CLS = BaseTypedReqSchema
+    CLS_ATTR_COOKING = attr.s(slots=True, frozen=True, auto_attribs=True, kw_only=True, eq=False)
+    DISCRIMINANT_FIELD = "cmd"
+
+    def __new__(  # type: ignore[no-untyped-def, misc]
+        cls, name: str, bases: Tuple[type, ...], nmspc: Dict[str, Any]
+    ):
+        # Sanity checks
+        if "SCHEMA_CLS" not in nmspc:
+            raise RuntimeError("Missing attribute `SCHEMA_CLS` in class definition")
+        if not issubclass(nmspc["SCHEMA_CLS"], cls.BASE_SCHEMA_CLS):
+            raise RuntimeError(f"Attribute `SCHEMA_CLS` must inherit {cls.BASE_SCHEMA_CLS!r}")
+
+        # Special case if we are defining `BaseReq/BaseRep`
+        if nmspc["SCHEMA_CLS"] is not cls.BASE_SCHEMA_CLS:
+            # Sanity check: discriminant field should be defined as check constant
+            discriminant_field = nmspc["SCHEMA_CLS"]._declared_fields.get(cls.DISCRIMINANT_FIELD)
+            if not discriminant_field or not isinstance(discriminant_field, fields.CheckedConstant):
+                raise RuntimeError(
+                    f"Attribute `SCHEMA_CLS` must contain a `{cls.DISCRIMINANT_FIELD}` field"
+                )
+
+        raw_cls = type.__new__(cls, name, bases, nmspc)
+
+        # Sanity checks: class SCHEMA_CLS needs to define parents SCHEMA_CLS fields
+        schema_cls_fields = set(nmspc["SCHEMA_CLS"]._declared_fields)
+        bases_schema_cls = (base for base in bases if hasattr(base, "SCHEMA_CLS"))
+        for base in bases_schema_cls:
+            assert (
+                base.SCHEMA_CLS._declared_fields.keys()  # type: ignore[attr-defined]
+                <= schema_cls_fields
+            )
+
+        # Sanity check: attr fields need to be defined in SCHEMA_CLS
+        if "__attrs_attrs__" in nmspc:
+            assert {att.name for att in nmspc["__attrs_attrs__"]} <= schema_cls_fields
+
+        raw_cls.SERIALIZER = MsgpackSerializer(  # type: ignore[attr-defined]
+            nmspc["SCHEMA_CLS"], InvalidMessageError, MessageSerializationError
+        )
+
+        # Define attribute `BaseTypedReqSchema.TYPE`
+        raw_cls.SCHEMA_CLS.TYPE = raw_cls  # type: ignore[attr-defined]
+
+        return raw_cls
+
+
+class CmdRepMeta(CmdReqMeta):
```

Comment:
So a `CmdRep` is also a `CmdReq`?

Reply:
https://github.com/Scille/parsec-cloud/pull/2155#issuecomment-1100069985

Reply:
807de9b7360674780a7edb4db70f3203b335aada